action/protocol/staking/candidate_center_extra.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2022 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+func (cb *candBase) clone() *candBase {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	clone := newCandBase()
+	for name, cand := range cb.nameMap {
+		clone.nameMap[name] = cand.Clone()
+	}
+	for owner, cand := range cb.ownerMap {
+		clone.ownerMap[owner] = cand.Clone()
+	}
+	for operator, cand := range cb.operatorMap {
+		clone.operatorMap[operator] = cand.Clone()
+	}
+	for bucket, cand := range cb.selfStkBucketMap {
+		clone.selfStkBucketMap[bucket] = cand.Clone()
+	}
+	if len(cb.owners) > 0 {
+		for _, cand := range cb.owners {
+			clone.owners = append(clone.owners, cand.Clone())
+		}
+	}
+	return clone
+}
+
+func (cb *candBase) candsInNameMap() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	if len(cb.nameMap) == 0 {
+		return nil
+	}
+
+	list := make(CandidateList, 0, len(cb.nameMap))
+	for _, d := range cb.nameMap {
+		list = append(list, d.Clone())
+	}
+	return list
+}
+
+func (cb *candBase) candsInOperatorMap() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	if len(cb.operatorMap) == 0 {
+		return nil
+	}
+
+	list := make(CandidateList, 0, len(cb.operatorMap))
+	for _, d := range cb.operatorMap {
+		list = append(list, d.Clone())
+	}
+	return list
+}
+
+func (cb *candBase) ownersList() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	return cb.owners
```

Comment:
for read, if we do clone in https://github.com/iotexproject/iotex-core/pull/3661/files#diff-17d187e581661db50bb29a2354c11c2d8e05245a5d4ad4a27f9619f6affb6ee7R56, https://github.com/iotexproject/iotex-core/pull/3661/files#diff-17d187e581661db50bb29a2354c11c2d8e05245a5d4ad4a27f9619f6affb6ee7R42, what about here?

Reply:
`name, op := candsInNameMap(), candsInOperatorMap()`
name, op will be serialized and saved to DB, so actually no need to `clone()` for the 2 places you mentioned, `clone()` is for safety I believe since we don't want to touch the `*Candidate` in these 2 maps.
for `ownerList`, the only use of it is to serialize and save to DB, so I think no need to clone the list here (b/c when we add a `*Candidate` to it, we make a clone)